de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java

```diff
@@ -532,44 +550,104 @@ public void addResourcesToSession(
             return;
         }
 
-        List<IProject> projectsToShare = new ArrayList<IProject>();
+        /** add new shared resources to a combining map for the next negotiation **/
+        synchronized (this) {
+            for (Entry<IProject, List<IResource>> mapEntry : projectResourcesMapping
+                .entrySet()) {
+                final IProject project = mapEntry.getKey();
```

Comment:
if (!containsKey(project) map.put(project, new Set<>)

if (resourceList == null || resourceList.isEmpty) {map.put(project, fullSharedMarkerSet); return;}

Set set = map.get(project)

if set ==  fullSharedMarkerSet
  return;

set.addAll(resourcesList)

Should be easier to understand.

Reply:
Please take in account that a new empty Set creates a partial project without resources.